stdlib/public/Differentiation/OptionalDifferentiation.swift

```diff
@@ -0,0 +1,83 @@
+//===--- OptionalDifferentiation.swift ------------------------*- swift -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Swift
+
+extension Optional: Differentiable where Wrapped: Differentiable {
+  public struct TangentVector: Differentiable, AdditiveArithmetic {
+    public typealias TangentVector = Self
+
+    public var value: Wrapped.TangentVector?
+
+    public init(_ value: Wrapped.TangentVector?) {
+      self.value = value
+    }
+
+    public static var zero: Self {
+      return Self(.zero)
```

Comment:
Note: it seems more efficient to use `Optional.TangentVector(nil)` as `Optional.TangentVector.zero`. We can explore the implications when working on differentiation support for `Optional`.